src/mlpack/methods/ann/layer/virtual_batch_norm.hpp

```diff
@@ -0,0 +1,194 @@
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+  typename InputDataType = arma::mat,
+  typename OutputDataType = arma::mat
+>
+class VirtualBatchNorm
+{
+ public:
+  //! Create the VirtualBatchNorm object.
+  VirtualBatchNorm();
+
+  /**
+   * Create the VirtualBatchNorm layer object for a specified number of input units.
+   *
+   * @param referenceBatch the data from which the normalization
+   *        statistics are computed.
+   * @param size The number of input units.
+   * @param eps The epsilon added to variance to ensure numerical stability.
+   */
+  template<typename eT>
+  VirtualBatchNorm(const arma::Mat<eT>& referenceBatch,
+                   const size_t size,
+                   const double eps = 1e-8);
+
+  /**
+   * Reset the layer parameters
+   */
+  void Reset();
+
+  /**
+   * Forward pass of the Virtual Batch Normalization layer. Transforms the input data
+   * into zero mean and unit variance, scales the data by a factor gamma and
+   * shifts it by beta.
+   *
+   * @param input Input data for the layer
+   * @param output Resulting output activations.
+   */
+  template<typename eT>
+  void Forward(const arma::Mat<eT>&& input, arma::Mat<eT>&& output);
+
+  /**
+   * Backward pass through the layer.
+   *
+   * @param input The input activations
+   * @param gy The backpropagated error.
+   * @param g The calculated gradient.
+   */
+  template<typename eT>
+  void Backward(const arma::Mat<eT>&& /* input */,
+                arma::Mat<eT>&& gy,
+                arma::Mat<eT>&& g);
+
+  /**
+   * Calculate the gradient using the output delta and the input activations.
+   *
+   * @param input The input activations
+   * @param error The calculated error
+   * @param gradient The calculated gradient.
+   */
+  template<typename eT>
+  void Gradient(const arma::Mat<eT>&& /* input */,
+                arma::Mat<eT>&& error,
+                arma::Mat<eT>&& gradient);
+
+  //! Get the parameters.
+  OutputDataType const& Parameters() const { return weights; }
+  //! Modify the parameters.
+  OutputDataType& Parameters() { return weights; }
+
+  //! Get the output parameter.
+  OutputDataType const& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the gradient.
+  OutputDataType const& Gradient() const { return gradient; }
+  //! Modify the gradient.
+  OutputDataType& Gradient() { return gradient; }
+
+
+  /**
+   * Serialize the layer
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  //! Locally-stored number of input units.
+  size_t size;
+
+  //! Locally-stored epsilon value.
+  double eps;
+
+  //! Whether we are in loading or saving mode.
```

Comment:
I think this can be something like `Variable to keep track of whether we are in ...`

Reply:
we are using the same description in `LayerNorm` and `BatchNorm`

Reply:
Then they should be changed as well :)